### DISCUSSION TITLE
Generate execution flow diagram

### DISCUSSION THREAD

**Discussion by markerikson** [RecordReplay/devtools#5924] [Ideas]
We're recording the entire app.  Can we generate an execution flow diagram from that info?

I'd love to be able to see, say, a trace of the startup sequence for Replay itself:

- Recording page component
- Redux store gets created somewhere
- Extra store bootstrap code runs
- Some client data gets fetched
- etc

Like, give me a way to visualize what files I'm jumping through, including the ability to blackbox things like React so we skip that stuff.   Some potential visualizations:

- Filenames / line snippets
- some kind of `dotviz` graph
- UML sequence diagrams
- gotta be plenty of other examples out there

**Comment by jasonLaster** [comment]
@markerikson how much of this would be solved by showing a call tree and user timings? curious what use cases would not be captured...

**Reply by jasonLaster** [reply]
Also - i know that a lot of user timings are dev only, but we can do a lot of this similar stuff at the framework level after the fact

**Reply by markerikson** [reply]
A call tree is probably one valid way to visualize this kind of info.  I'm also sorta picturing (*waves hands wildly*) something more diagram-y, possibly with embedded snippets of code kinda like we have in comments.

**Comment by markerikson** [comment]
From @kettanaito in Reactiflux, asking separately without knowing about Replay :

> Is there a concept opposite to Error stack traces?
Stack trace prints the execution stack before the error. I'd like to know the execution stack after a certain point. 
> 
>Let's say I import a foo function from a third-party package. I want to go through all the underlying code it executes. foo is quite complex, and I don't even know where exactly it goes with its logic. Is there a way to print the execution "path"?
>
> In other words, I'd like to travel through foo as it executes on runtime. So the execution is rather precise, fixed (foo receives concrete parameters that lead it to a concrete path in a static source code). There must be something for this.

**Reply by kettanaito** [reply]
I was excited to find out Replay can do something similar to what I'm looking for. My work is focused on creating tools that work in harmony with other tools. For that, I often inspect how things work. Being able to achieve a tangible execution trace would be awesome.

**Comment by markerikson** [comment]
Some more ideas and feedback, including suggesting using a "flame graph" diagram for navigating the execution of a replay:

- https://twitter.com/bitandbang/status/1507775758495891459
- https://twitter.com/geoffreylitt/status/1507777464868741121

**Reply by markerikson** [reply]
The `dbux` debugger tool generates some async call flow diagrams:

https://domiii.github.io/dbux/acg

Pasting some screenshots:

![image](https://user-images.githubusercontent.com/1128784/160720597-6b1fc594-b9d1-4b70-a358-31f71b1d3790.png)

![image](https://user-images.githubusercontent.com/1128784/160720646-fdc7c8ea-6712-4636-810b-e0ef23b4f011.png)

Those look basically like what I had in my head

**Comment by Domiii** [comment]
hi, this is Domi, creator of [Dbux](https://github.com/Domiii/dbux/).

Let me quickly introduce Dbux's "maps", i.e. global runtime visualization tools:

(But I definitely also recommend checking out the much more complete (and generously timestamped) [intro video](https://www.youtube.com/watch?v=N9W6rhHMKbA).)

1. The [Synchronous Call Graph](https://domiii.github.io/dbux/runtime-analysis/call-graph) is a basic dynamic call graph.
   * It is similar to the flamegraph, but designed for investigating control flow, and, unlike the flamegraph, not optimized for profiling.
   * It allows you to not only see the "call stack evolved over time", but also see inputs and outputs of calls etc.
   * [Call Graph Roots](https://domiii.github.io/dbux/acg#cgr) are files/functions (or partial files/functions) that are called or resumed from the JS engine, i.e. that do not have a caller.
   * The roots are just listed on a single (vertical) timeline.
   * The dynamic call graph has been implemented many times before, but only either for (1) profiling, or (2) in research experiments. To my knowledge, there is no proper interactive call graph tool for control flow analysis out there. But please let me know, if you have seen anything even remotely close!
3. The [Asynchronous Call Graph](https://domiii.github.io/dbux/acg), on the other hand, is our own invention.
   * Our (short) paper on the ACG just got accepted to [ICSE 2022](https://conf.researchr.org/track/icse-2022/icse-2022-seip---software-engineering-in-practice) (link not up yet).
   * Asynchronous execution models are concurrent execution models. The ACG's goal is to capture that concurrency and, in doing so, compliment the synchronous call graph (above). It is designed to help us understand the relationship at the level where there are no callers, that is only between the roots.
   * The `val` (value) button allows you to render the value of a selected expression for every root, allowing you to identify which root handled which data.
4. The [Asynchronous Call Stack](https://domiii.github.io/dbux/runtime-analysis/call-graph#stack) stitches the data used for the two of them to produce the complete asynchronous call stack at point of investigation. Turns out, there is no JavaScript environment that produces a complete asynchronous call stack that can cope with all different types of [asynchronous events](https://domiii.github.io/dbux/acg#ae).

## Some realizations after starting to use these tools

1. I don't want to live without them anymore.
2. They changed my entire perception and ability to reason about the overall call graph structure.
3. Because of (2), even, when I don't have them available, I feel I am in a much better position to cull my search space, once I understand some of the caller-callee relationships.
5. The two graphs compliment each other. You would usually want both. Without the ACG, a lot of the concurrency escapes you. Without the SCG, the synchronous call stack fragments might elude you. But maybe, just having the ACG + the Asynchronous Call Stack (ACS) would suffice.
6. Once you start using these tools to hunt for bugs, you thirst for more. I have already started adding more little extras to help us answer all kinds of remaining questions:
   * Given a promise, what are all its own and its nested roots (the paper's technical report explains the semantics of CGR and promise nesting)?
   * Given any root, promise or async function invocation, which line of code scheduled it?
   * Given any root, promise or async function invocation, at a specific point in time, what was the last piece of code it executed before this root?
   * Given any root, promise or async function invocation, what is it synchroizing against? (Promises sometimes have their own asynchronous control flow (i.e. their own roots), but sometimes, they are used to synchronize between control flows of other promises.)
   * And many more!


## Implementation + Performance notes

1. Naturally, you would always have a lot more file/function executions than roots.
   * Thus, the SCG can get pretty heavy.
   * You probably want to build it partially/on the fly, if you can. Dbux is doing Ok with up to several hundred thousand calls.
   * You probably want to be able to filter packages and/or file path regexes to cull the search space. Dbux has that, too.
     * Having all libraries in there can make things unnecessarily messy.
2. The data needed for the ACG, on the other hand is usually relatively small.
   * I have not done the analysis on this yet. But naturally, you won't usually have more than a few hundred (or worst case, a few thousand) call graph roots.
   * Unless you have long running `async` loops. In that case, some loop comprehension/summarization probably can reduce the data needed there.

## Some more pretty pictures and stuff

###  Dbux (recent) Intro Video
![Video Introduction](https://img.youtube.com/vi/N9W6rhHMKbA/0.jpg)

### Fibonacci Call Graph vs. Call Stack (w/ `val` turned on)
![](https://domiii.github.io/dbux/screens/sample-fibonacci-graph-vs-stack.png)

### Conceptual ACG demonstrating how missing `await`s lead to FORKing
![](https://domiii.github.io/dbux/concepts/chain_fork_3.png)


### Sequelize ACG: running two queries in parallel (zoomed out, details disabled)
* NOTE1: Somewhere near the middle, you see the two queries forking off into two individual "threads".
* NOTE2: If an Error is thrown inside of a node, it is visualized with a flame icon.
  * side note: Those error icons helped me find a logged bug and a second, not logged, bug basically instantly (when you click the flame icon, they take you to the origin of the error).

![sequelize5](https://user-images.githubusercontent.com/282899/160816358-ca51e0c2-429d-4c49-be25-1e02a60d8169.png)

### Editor.md ACG: editing and saving markdown strings.
* The two long purple "threads" in the middle are UI event handler invocations. I think it was `keyup` and `keydown` events.
* The "beige" roots forking off to the right are buffered auto-save events

![Editor md](https://user-images.githubusercontent.com/282899/160816382-af3040f7-59c1-4a83-b631-23a026437a28.png)


## In Conclusion

If done right, these graphs can tell powerful stories.

Let me know, if there is anything I can do to help!

**Reply by markerikson** [reply]
YES YES YES!  This is _ABSOLUTELY_ the kind of visualization I want to see in Replay!

**Comment by markerikson** [comment]
Another set of interesting diagrams: Kolo for Django:

https://kolo.app/

![image](https://user-images.githubusercontent.com/1128784/161439337-d545bded-4725-4a9a-a834-f8ccef915a94.png)

**Reply by Domiii** [reply]
Reminds me of the "circular bundle view" from `Extravis` ([Execution Trace Analysis through Massive Sequence and Circular Bundle Views](https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.212.9450&rep=rep1&type=pdf), 2008). Even though, they work very differently, they both represent circular call graph summaries.
![image](https://user-images.githubusercontent.com/282899/163027788-4e64faa6-4f3e-410b-8f27-b1ab2ebc7d0d.png)

I guess, the `Kolo` visualization can quickly get out of control, when you have more than a few events (I'm guessing, I see about 10-ish what-I-guess is routes (or also other types of events?) in the Django screenshot?); but then again, everything will spin out of control, if you just throw enough data at it - hah!

If I'm not mistunderstanding them, the `Extravis` authors claim such "summary views" can be useful for some overview tasks, but usually need some complimenting views to better orient yourself and detect patterns-over-time.

**Comment by markerikson** [comment]
Some more prior art:

https://github.com/janestreet/magic-trace

![magic-trace UI](https://github.com/janestreet/magic-trace/raw/master/docs/assets/stage-3.gif)

**Reply by Domiii** [reply]
Not to be a negative Nancy, but is this different from Browser Dev Tools' profiler's flamecharts or icycle plots?

**Reply by Domiii** [reply]
I personally always prefer putting the time dimension on "control flow maps" on the vertical dimension (and of course only if there are two clear dimensions to begin with), if the goal is readability and comprehension (rather than, in case of profilers: highlighting of bottlenecks).
Else the text gets very hard to read.

Generally speaking: if the goal is "highlighting of patterns", of course more discussions are to be had, but generally, I would avoid trying to stack text horizontally.

**Comment by markerikson** [comment]
Throwing something else in here:

This extremely long article on perf profiling the guts of the Rust compiler talks about [OpenTelemetry](https://opentelemetry.io/), and how to generate "spans" that can be visualized.  Maybe something related we could do?

https://fasterthanli.me/articles/when-rustc-explodes

![image](https://user-images.githubusercontent.com/1128784/179074113-2547160a-a1b5-4a76-b324-4f551a62ba42.png)

**Reply by Domiii** [reply]
Is the goal to aid profiling or program comprehension?

**Reply by markerikson** [reply]
Comprehension, mostly.  If there's a crash with an error, I'd like to be able to see not just "the stack trace at the time of the error", but also "what functions and logic did we execute leading up to this spot".

For example, when I was debugging https://app.replay.io/recording/logpoint-02js--2d7e7544-33f3-4e39-b82d-b4f7d0b2daf5?point=22067261651551680457036465793139134&time=10896.647887323943&hasFrames=true# yesterday, I had to do a bunch of back-tracing to go from the final test failure to figure out where things actually went wrong long before that.  Having this sort of a trace diagram might have made that faster.

**Reply by Domiii** [reply]
It looks like a normal, vertically arranged [call graph](https://domiii.github.io/dbux/dynamic-analysis/call-graph)  but a bit coarser, focusing only on processes and system interactions?
